contracts/Market.sol

```diff
@@ -1929,6 +1771,168 @@ contract Market is
     return (countOfRemovalsAllocated, ids, amounts);
   }
 
+  /**
+   * @notice Allocates the removals, amounts, and suppliers needed to fulfill the purchase drawing only from
+   * the vintages specified.
+   * @param amount The number of carbon removals to purchase.
+   * @param vintages The vintages from which to purchase carbon removals.
+   * @return countOfRemovalsAllocated The number of distinct removal IDs used to fulfill this order.
+   * @return ids An array of the removal IDs being drawn from to fulfill this order.
+   * @return amounts An array of amounts being allocated from each corresponding removal token.
+   * @return suppliers The address of the supplier who owns each corresponding removal token.
+   */
+  function _allocateSupplySpecificVintages(
+    uint256 amount,
+    uint256[] memory vintages
+  )
+    private
+    returns (
+      uint256 countOfRemovalsAllocated,
+      uint256[] memory ids,
+      uint256[] memory amounts,
+      address[] memory suppliers
+    )
+  {
+    uint256 remainingAmountToFill = amount;
+    uint256 countOfListedRemovals = _removal.numberOfTokensOwnedByAddress({
+      account: address(this)
+    });
+    ids = new uint256[](countOfListedRemovals);
+    amounts = new uint256[](countOfListedRemovals);
+    suppliers = new address[](countOfListedRemovals);
+    countOfRemovalsAllocated = 0;
+    for (uint256 i = 0; i < countOfListedRemovals; ++i) {
+      uint256 removalId = _listedSupply[_currentSupplierAddress]
+        .getNextRemovalForSaleFromVintages({vintages: vintages});
+      // if removalId is 0 here, this supplier doesn't have any removals from the vintages requested
```

Comment:
this comment is actually part of the natspec for that function so i'm going to delete this comment that was just part of my pseudocoding

Reply:
Still could be useful to refactor to an internal function since there's a lot happening here

Reply:
the only other activity here is setting the `localCurrentSupplier` to the next in the case that this supplier is out of removals... adding another internal function will add useless overhead and leave the same number of lines here because we still have to assign this local variable

Reply:
Oh, maybe I'm looking at a stale diff? This diff looks like several variables are updated